### ISSUE TITLE
Feature request: SVG graph rendering option

### ISSUE THREAD

**Issue by vituperative** [rrd4j/rrd4j#165]

**Comment by fbacchella** [comment]
Rendering is done using java AWT  and uses ImageIO. Adding SVG support is not straight forward, but if you submit a PR, I will merge it.

**Comment by vituperative** [comment]
Would this library simplify things? https://github.com/jfree/jfreesvg

**Comment by vituperative** [comment]
I looked at twelvemonkeys and noticed the apache batik dependency which is why I didn't suggest it, given the size of the batik lib. I'll have a look at your new branch, though, thanks!

**Comment by fbacchella** [comment]
twelvemonkeys uses the ServiceProvider API. So just dropping the needed jars and using svg as the file format is enough.

**Comment by zzzi2p** [comment]
OP and I got your patch working with jfreesvg, thanks.

My only suggestion is that you make all the ImageWorker methods public or protected, so we don't have to put our extending class into your package.

**Comment by zzzi2p** [comment]
We're staring at the generated SVGs and see some things of interest that would not be apparent for other image formats.

First, the path for an AREA is in there twice, first as a fillPolygon and then as a drawPolygon. This almost doubles the size of the SVG.

It comes from ImageWorker.fillPolygon(double[], double, double[], Paint) which calls both d2g.fillPolygon() and d2g.drawPolygon(). While we can easily override the method to remove the d2g.drawPolygon() call, we question why it's in there to begin with. There's no color change between the two calls and it's using the default BasicStroke(1).  And there's no drawPolygon() in the related method below it, which is used for STACK.

The only way to see the effects of drawPolygon() at all is if the area color alpha is significantly less than 255, because it will get two layers written for the border of the area and it will be a little darker, but the effect is extremely subtle.

So we recommend that the g2d.drawPolygon() call be removed, unless you're going to change the color before calling it.

We will probably follow up with some other issues we're researching now.

**Comment by zzzi2p** [comment]
Second thing, the polygon paths contain a large number of duplicate points, presumably due to truncation from double to int. This is inefficient in any rendering format, but is only apparent for SVG where it bloats the size of the file.

The following patch (against master, not the custom branch, but should apply to either one) dedups the points before passing to g2d, for area, stack, and line. It also comments out the drawPolygon() call referenced above.

```diff
diff --git a/src/main/java/org/rrd4j/graph/ImageWorker.java b/src/main/java/org/rrd4j/graph/ImageWorker.java
index 1c5aca6..e6ab701 100644
--- a/src/main/java/org/rrd4j/graph/ImageWorker.java
+++ b/src/main/java/org/rrd4j/graph/ImageWorker.java
@@ -52,15 +52,24 @@ public abstract class ImageWorker {
         for (int[] pos = path.getNextPath(); pos != null; pos = path.getNextPath()) {
             int start = pos[0], end = pos[1], n = end - start;
             int[] xDev = new int[n + 2], yDev = new int[n + 2];
+            int c = 0;
+            int px = -999;
+            int py = -999;
             for (int i = start; i < end; i++) {
-                xDev[i - start] = (int) x[i];
-                yDev[i - start] = (int) yTop[i];
+                int cx = (int) x[i];
+                int cy = (int) yTop[i];
+                if (cx != px || cy != py) {
+                    xDev[c] = cx;
+                    yDev[c++] = cy;
+                    px = cx;
+                    py = cy;
+                }
             }
-            xDev[n] = xDev[n - 1];
-            xDev[n + 1] = xDev[0];
-            yDev[n] = yDev[n + 1] = (int) yBottom;
-            g2d.fillPolygon(xDev, yDev, xDev.length);
-            g2d.drawPolygon(xDev, yDev, xDev.length);
+            xDev[c] = xDev[c - 1];
+            xDev[c + 1] = xDev[0];
+            yDev[c] = yDev[c + 1] = (int) yBottom;
+            g2d.fillPolygon(xDev, yDev, c + 2);
+            //g2d.drawPolygon(xDev, yDev, c + 2);
         }
     }
 
@@ -70,13 +79,30 @@ public abstract class ImageWorker {
         for (int[] pos = path.getNextPath(); pos != null; pos = path.getNextPath()) {
             int start = pos[0], end = pos[1], n = end - start;
             int[] xDev = new int[n * 2], yDev = new int[n * 2];
+            int c = 0;
+            int px = -999;
+            int py = -999;
             for (int i = start; i < end; i++) {
-                int ix1 = i - start, ix2 = n * 2 - 1 - i + start;
-                xDev[ix1] = xDev[ix2] = (int) x[i];
-                yDev[ix1] = (int) yTop[i];
-                yDev[ix2] = (int) yBottom[i];
+                int cx = (int) x[i];
+                int cy = (int) yTop[i];
+                if (cx != px || cy != py) {
+                    xDev[c] = cx;
+                    yDev[c++] = cy;
+                    px = cx;
+                    py = cy;
+                }
             }
-            g2d.fillPolygon(xDev, yDev, xDev.length);
+            for (int i = end - 1; i >= start; i--) {
+                int cx = (int) x[i];
+                int cy = (int) yBottom[i];
+                if (cx != px || cy != py) {
+                    xDev[c] = cx;
+                    yDev[c++] = cy;
+                    px = cx;
+                    py = cy;
+                }
+            }
+            g2d.fillPolygon(xDev, yDev, c);
         }
     }
 
@@ -93,11 +119,20 @@ public abstract class ImageWorker {
         for (int[] pos = path.getNextPath(); pos != null; pos = path.getNextPath()) {
             int start = pos[0], end = pos[1];
             int[] xDev = new int[end - start], yDev = new int[end - start];
+            int c = 0;
+            int px = -999;
+            int py = -999;
             for (int i = start; i < end; i++) {
-                xDev[i - start] = (int) x[i];
-                yDev[i - start] = (int) y[i];
+                int cx = (int) x[i];
+                int cy = (int) y[i];
+                if (cx != px || cy != py) {
+                    xDev[c] = cx;
+                    yDev[c++] = cy;
+                    px = cx;
+                    py = cy;
+                }
             }
-            g2d.drawPolyline(xDev, yDev, xDev.length);
+            g2d.drawPolyline(xDev, yDev, c);
         }
     }
 ```

**Comment by zzzi2p** [comment]
and yes it could be cleaned up to look at the previous values in the arrays, and get rid of the px/py variables, if that's more to your liking

there's another optimization to be had, to collapse straight lines, if the previous two points have the same Y as the current point, replace the previous point with the current point

**Comment by fbacchella** [comment]
If you provide a PR, I will merge it.

**Comment by zzzi2p** [comment]
OK I'll work on the cleanups I mentioned above first. You want it against master or the custom branch? You have any schedule for the next release?

**Comment by fbacchella** [comment]
Let’s do it again the custom branch. There is nothing scheduled but I will do a release soon atfter the merge.

**Comment by zzzi2p** [comment]
Thanks for all your help. I put up our ImageWorker extension for anybody who wants it at https://github.com/zzzi2p/rrd4j-jfreesvg

**Comment by fbacchella** [comment]
I ran your code on TestDemo and got strange results, the legends are written vertically. I’m not fluent with SVG, so I wonder what happens.
I created a branch TestSvg with your image worker to check it. Can you have a look ? If you run `mvn test -Dtest=TestDemo` it will generate a /tmp/test.png  file.

**Comment by zzzi2p** [comment]
Reproduced here. We're not resetting the transform in SVGImageWorker.reset() after you draw the vertical label, like you do in BufferedImageWorker. Will fix, retest, and report back.

**Comment by zzzi2p** [comment]
Fixed in https://github.com/zzzi2p/rrd4j-jfreesvg/commit/e274fbaf6a56cdadd1488ff53b5dbc65ac556e30
Verified with TestDemo
Thanks for writing a good test!

**Comment by fbacchella** [comment]
It works, nice. Another question, the used font (DejaVu Sans Mono) was not installed on my machine and the font name was explicitely required in the SVG.
It generate:

```
<text x="0.0" y="0.0" style="fill: rgb(192,192,192); fill-opacity: 1.0; font-family: &quot;DejaVu Sans Mono&quot;; font-size: 9px;" text-rendering="optimizeLegibility" clip-path="url(#_159744825767375clip-4)">Generated by RRD4J</text>
```

I wonder if JFeeSVG provides any way to handle that ?

**Comment by zzzi2p** [comment]
Yes, it works, verified in our usage. ImageWorker.drawString() calls g2d.setFont(). You just need to call gDef.setFont() in TestDemo if you want something different. It looks like the jfreesvg default is SansSerif 12 pt, at least in the version I have, not sure where DejaVu is coming from for you.

**Comment by zzzi2p** [comment]
Actually it comes out DejaVu for me also. I think somewhere deep in the Java AWT font stuff it maps SansSerif to DejaVu Sans Mono.

**Comment by fbacchella** [comment]
That's the default font for rrd4j, embeded in the library, to generate standard rendering. Buf for vector rendering, that's no the best option.

**Comment by zzzi2p** [comment]
Batik version working, briefly tested only, checked in same repo as link above. No need for TwelveMonkeys. Pros: Apache license, much bigger and more flexible API for doing advanced things, more optimized SVG output, seems to be better maintained; Cons: minimal dependency is 8 of the ~35 Batik jars, totaling about 1 MB (vs. 50 KB for jfreesvg)

**Comment by fbacchella** [comment]
I merged you sample code as a test if you don’t mind. If you are ok, I will release a version 3.10

**Comment by zzzi2p** [comment]
That's fine, same license as yours. We don't have anything more coming from our side. Thanks for your help.

**Comment by fbacchella** [comment]
I’m finishing the last tweaks, improving the handling of log scale, and it’s almost done.
You can create create the pull request. I will have a look at it and probably merge it.

**Comment by fbacchella** [comment]
Integrated in RRD4J 3.10
